### PR TITLE
Conversation tests using yaml

### PR DESCRIPTION
**Pull request by tmbo**
**Proposed changes**:
- always ensure `UserUttered` events contain the full `parse_data`
- renamed occurrences of `conversation_tests.md` to `test_conversations.yml` (as well as remaining `nlu.md` and `story.md` ones)
- added default tests to `rasa init` in yaml format
- removed the need to pass an interpreter to story readers (wasn't actually needed, e2e evaluation is done slightly different instead)
- some code style improvements here and there 😅  (sorry for that, lead to an unexpected amount of changes)

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by tmbo** [comment]
@erohmensing I've updated the documentation to reflect the changes to the story testing parts. Also updated the chitchat-faq's page along the way. (and some other smaller things I noticed in the docs). do you have time to review the docs changes?

File: docs/docs/command-line-interface.mdx [hunk 1]
```diff
@@ -25,7 +25,7 @@ The command line interface (CLI) gives you easy-to-remember commands for common
 |`rasa x`                |Launch Rasa X locally.                                                                                                                    |
 |`rasa -h`               |Shows all available commands.                                                                                                             |
 
-## Create a new project
+## rasa init
```
Comment (tmbo) [inline docs/docs/command-line-interface.mdx:28]: I changed the headlines as it makes it a lot easier to find and reference the commands. I think it makes more sense from a users perspective as this page is meant to look up commands.

File: rasa/core/events/__init__.py [hunk 2]
```diff
@@ -56,18 +56,16 @@ def deserialise_entities(entities: Union[Text, List[Any]]) -> List[Dict[Text, An
...
-    message_from_md = MarkdownReader().parse_training_example(text)
+    message_from_md = entities_parser.parse_training_example(text, intent)
     deserialised_entities = deserialise_entities(entities)
     return TrainingDataWriter.generate_message(
-        {
-            "text": message_from_md.text,
-            "intent": intent,
-            "entities": deserialised_entities,
-        }
+        {"text": message_from_md.text, "entities": deserialised_entities,}
```
Comment (tmbo) [inline rasa/core/events/__init__.py:68]: `intent` wasn't used here at all

File: docs/docs/command-line-interface.mdx (hunk 1 above)
Comment (erohmensing) [reply docs/docs/command-line-interface.mdx:28]: I 100% agree!

File: rasa/utils/plotting.py [hunk 3]
```diff
@@ -12,18 +12,34 @@
...
-    pass
-else:  # pragma: no cover
-    try:
-        # If the `tkinter` package is available, we can use the `TkAgg` backend
-        import tkinter
-
-        matplotlib.use("TkAgg")
-    except ImportError:
-        matplotlib.use("agg")
+def _fix_matplotlib_backend() -> None:
```
Comment (tmbo) [inline rasa/utils/plotting.py:15]: fix is necessary due to issues of macos with tkiner on specific osx versions - the previous implementation was not working for me and matplot lib code failed with the tkagg error. this fixes that issue.

**Comment by tmbo** [comment]
@degiz this is ready for a code review, still need to add some tests but I think giving the code a review first helps speed this up.

File: docs/docs/chitchat-faqs.mdx [hunk 4]
```diff
@@ -145,6 +107,58 @@ nlu:
     - cheers
 ```
 
+### 2. Writing stories
```
Comment (erohmensing) [inline docs/docs/chitchat-faqs.mdx:110]: This entire page will be rewritten/restructured here: https://github.com/RasaHQ/rasa/pull/6373
Wouldn't recommend putting too much work into it 😃 (as well as all of the conversations pattern docs)

File: docs/docs/command-line-interface.mdx [hunk 5]
```diff
@@ -272,7 +298,7 @@ should be published.
 ```text [rasa export --help]
 ```
 
-## Start Rasa X
+## rasa x
 
 Rasa X is a toolset that helps you leverage conversations to improve your assistant.
 You can find more information about it <a className="reference external" href="https://rasa.com/docs/rasa-x/" target="_blank">here</a>.You can start Rasa X locally by executing
```
Comment (erohmensing) [inline docs/docs/command-line-interface.mdx:304]: looks like the :::note after this did not make it through the mirgration, would you mind fixing that?

File: docs/docs/testing-your-assistant.mdx [hunk 6]
```diff
@@ -162,24 +158,23 @@ into train and test sets using:
...
+rasa test nlu --nlu train_test_split/test_data.yml
 ```
 
 If you don't want to create a separate test set, you can
 still estimate how well your model generalises using cross-validation.
 To do this, add the flag `--cross-validation`:
 
 ```bash
-rasa test nlu -u data/nlu.md --config config.yml --cross-validation
+rasa test nlu --nlu data/nlu.ml --cross-validation
```
Comment (erohmensing) [inline docs/docs/testing-your-assistant.mdx:173]:
```suggestion
rasa test nlu --nlu data/nlu.yml --cross-validation
```

File: docs/docs/training-data-format.mdx [hunk 7]
```diff
@@ -71,21 +83,13 @@ rules:
...
-    - hey there [Sara](name)
-
-- intent: faq/language
-  examples: |
-    - What language do you speak?
-    - Do you only handle english?
+```
 
-e2e_tests:
+If you want to specify your test stories, you need to put them into a separate file:
```
Comment (erohmensing) [inline docs/docs/training-data-format.mdx:88]: Should we mention that this file's name should start with `test_`?

File: rasa/core/events/__init__.py (hunk 2 above)
Comment (erohmensing) [reply rasa/core/events/__init__.py:68]: should this method still take it as an argument then?

File: rasa/core/test.py [hunk 8]
```diff
@@ -245,23 +257,21 @@ def _clean_entity_results(
...
     fail_on_prediction_errors: bool,
 ) -> EvaluationStore:
     user_uttered_eval_store = EvaluationStore()
 
-    intent_gold = event.parse_data.get("true_intent")
-    predicted_intent = event.parse_data.get("intent", {}).get("name")
-
-    if not predicted_intent:
-        predicted_intent = [None]
+    intent_gold = event.intent.get("name")
```
Comment (erohmensing) [inline rasa/core/test.py:266]: where was the `"true_intent"` originally coming from and why did we switch?

File: rasa/core/training/story_reader/markdown_story_reader.py [hunk 9]
```diff
@@ -193,25 +198,58 @@ async def _add_e2e_messages(self, e2e_messages: List[Text], line_num: int) -> No
...
         self.current_step_builder.add_user_messages(parsed_messages)
 
-    async def _parse_message(self, message: Text, line_num: int) -> UserUttered:
+    @staticmethod
+    def parse_e2e_message(line: Text) -> "Message":
+        f"""Parses an md list item line based on the current section type.
+
+        Matches expressions of the form `<intent>:<example>. For the
+        syntax of <example> see the Rasa docs on NLU training data:
+        {DOCS_BASE_URL}/nlu/training-data-format/#markdown-format"""
```
Comment (erohmensing) [inline rasa/core/training/story_reader/markdown_story_reader.py:215]:
```suggestion
        f"""Parses an md list item line based on the current section type.

        Matches expressions of the form `<intent>:<example>`. For the
        syntax of `<example>` see the Rasa docs on NLU training data:
        {DOCS_BASE_URL}/training-data-format/#markdown-format"""
```

File: rasa/core/training/story_reader/markdown_story_reader.py [hunk 10]
```diff
@@ -193,25 +198,58 @@ async def _add_e2e_messages(self, e2e_messages: List[Text], line_num: int) -> No
...
+        item_regex = re.compile(r"\s*" + form_group + r"([^{}]+?)({.*})*:\s*(.*)")
+        match = re.match(item_regex, line)
+
+        if not match:
+            raise ValueError(
+                "Encountered invalid end-to-end format for message "
+                "`{}`. Please visit the documentation page on "
+                "end-to-end testing at {}/user-guide/testing-your-assistant/"
+                "#end-to-end-testing/".format(line, DOCS_BASE_URL)
+            )
```
Comment (erohmensing) [inline rasa/core/training/story_reader/markdown_story_reader.py:232]: might make more sense in this case just to link to the stories training data format?

File: rasa/core/training/story_reader/markdown_story_reader.py [hunk 11]
```diff
@@ -257,6 +295,26 @@ def is_markdown_story_file(file_path: Union[Text, Path]) -> bool:
...
 
+    @staticmethod
+    def is_markdown_test_stories_file(file_path: Union[Text, Path]) -> bool:
+        """Checks if a file contains test stories.
+
+        Args:
+            file_path: Path of the file which should be checked.
+
+        Returns:
+            `True` if it's a file containing test stories, otherwise `False`.
```
Comment (erohmensing) [inline rasa/core/training/story_reader/markdown_story_reader.py:306]: what about if it starts with `DEFAULT_TEST_STORIES_FILE_PREFIX ` ?

Comment (erohmensing) [reply rasa/core/training/story_reader/markdown_story_reader.py:306]: logic seems to be different from yaml

File: rasa/data.py [hunk 12]
```diff
@@ -61,28 +85,28 @@ def get_core_nlu_directories(
...
+) -> List[Text]:
     """Recursively collects all training files from a list of paths.
 
     Args:
         paths: List of paths to training files or folders containing them.
 
     Returns:
-        Tuple of paths to story and NLU files.
+        paths of training data files.
     """
```
Comment (erohmensing) [inline rasa/data.py:107]: please add the new argument inc. what the options are or where to find those options

**Review by erohmensing — COMMENTED** [review]
Awesome that you were able to pull out that `interpreter` from so much stuff 😄

**Comment by erohmensing** [comment]
Oops, I didn't see that you just wanted my review on docs changes, sorry 😅

**Comment by tmbo** [comment]
No worries, always also happy about any code comments 🙂

File: rasa/core/test.py (hunk 8 above)
Comment (tmbo) [reply rasa/core/test.py:269]: the true intent was attached in the training data reader, it wasn't necessary (as the value was always set to the same value as the intent, which makes sense when reading the data). not sure who implemented it, but it was a workaround - data gets pushed through the evaluation slightly different separating loaded data from predictions.

File: rasa/core/training/story_reader/markdown_story_reader.py (hunk 10 above)
Comment (tmbo) [reply rasa/core/training/story_reader/markdown_story_reader.py:232]: well, the new page won't contain the markdown info anyways - will link to the legacy docs instead

File: rasa/core/training/story_reader/markdown_story_reader.py (hunk 11 above)
Comment (tmbo) [reply rasa/core/training/story_reader/markdown_story_reader.py:306]: yes that is intentional, the logic is the same as before - don't want to change this as it would break the existing logic

File: rasa/constants.py [hunk 13]
```diff
@@ -8,6 +8,7 @@
 DEFAULT_MODELS_PATH = "models"
 DEFAULT_DATA_PATH = "data"
 DEFAULT_E2E_TESTS_PATH = "tests"
+DEFAULT_TEST_STORIES_FILE_PREFIX = "test_"
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/constants.py:11]: yes good point

File: rasa/core/processor.py [hunk 14]
```diff
@@ -441,18 +438,33 @@ def _check_for_unseen_features(self, parse_data: Dict[Text, Any]) -> None:
     def _get_action(self, action_name) -> Optional[Action]:
         return self.domain.action_for_name(action_name, self.action_endpoint)
 
-    async def _parse_message(self, message, tracker: DialogueStateTracker = None):
+    async def parse_message(
+        self, message: UserMessage, tracker: Optional[DialogueStateTracker] = None
+    ) -> Dict[Text, Any]:
+        """Interprete the passed message using the NLU interpreter.
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/processor.py:444]: mhm not sure what you mean, it will use the regex intereptreter only for messages starting with `/`

File: rasa/core/training/interactive.py [hunk 15]
```diff
@@ -1233,8 +1225,7 @@ async def _correct_entities(
     )
 
     annotation = await _ask_questions(question, conversation_id, endpoint)
-    # noinspection PyProtectedMember
-    parse_annotated = MarkdownReader().parse_training_example(annotation)
+    parse_annotated = entities_parser.parse_training_example(annotation, intent=None)
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/training/interactive.py:1228]: mhm no it's not, there is no default

File: rasa/core/training/story_reader/markdown_story_reader.py [hunk 16]
```diff
@@ -193,25 +195,57 @@ async def _add_e2e_messages(self, e2e_messages: List[Text], line_num: int) -> No
...
+        # 3) Optional entities
+        # 4) The message text
+        form_group = fr"({FORM_PREFIX}\s*)*"
+        item_regex = re.compile(r"\s*" + form_group + r"([^{}]+?)({.*})*:\s*(.*)")
+        match = re.match(item_regex, line)
+
+        if not match:
+            raise ValueError(
+                "Encountered invalid test story format for message "
+                "`{}`. Please visit the documentation page on "
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/training/story_reader/markdown_story_reader.py:225]: we do need to throw an exception to bubble up the stack to the caller of the parser, a warning isn't enough (I only copied this over, but just checked)

File: rasa/data.py [hunk 17]
```diff
@@ -61,28 +110,29 @@ def get_core_nlu_directories(
...
     nlu_directory = _copy_files_to_new_dir(nlu_data_files)
 
     return story_directory, nlu_directory
 
 
-def get_core_nlu_files(
-    paths: Optional[Union[Text, List[Text]]]
-) -> Tuple[List[Text], List[Text]]:
+def get_data_files(
+    paths: Optional[Union[Text, List[Text]]], filter_property: Callable[[Text], bool]
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/data.py:123]: yes good point

File: rasa/data.py [hunk 18]
```diff
@@ -156,37 +199,33 @@ def is_story_file(file_path: Text) -> bool:
...
-
-    if YAMLStoryReader.is_yaml_story_file(file_path):
-        return True
-
     from rasa.core.training.story_reader.markdown_story_reader import (
         MarkdownStoryReader,
     )
 
-    return MarkdownStoryReader.is_markdown_story_file(file_path)
+    return YAMLStoryReader.is_yaml_story_file(
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/data.py:206]: I am not sure what you mean 😅  is there something we should change?

**Comment by tmbo** [comment]
good point about the naming, I'll make sure it is test_stories everywhere instead of test conversations. I'll keep some occurrences in the docs page though for discoverability

File: rasa/core/processor.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (degiz) [reply rasa/core/processor.py:444]: ok looks like I was unfocused closer to the end 😓

File: rasa/core/training/interactive.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (degiz) [reply rasa/core/training/interactive.py:1228]: Sorry, I meant that `parse_training_example()` is passing the `intent` to the `Message.build()`, and the `build()` has default value of None, so `parse_training_example()` can have it as well. Does it make sense?

File: rasa/core/training/story_reader/markdown_story_reader.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (degiz) [reply rasa/core/training/story_reader/markdown_story_reader.py:225]: True, got lost 🤦

File: rasa/data.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (degiz) [reply rasa/data.py:206]: I meant that both methods have exactly the same body now, maybe we can leave only one of them?

[the remark this answers is not in the document]
Comment (degiz) [reply rasa/data.py:206]: Ah no ignore 🤦

**Comment by tmbo** [comment]
@degiz addressed all the comments - still need an accept though 😉

**Review by degiz — APPROVED** [review]
Let's go 🚀 👍

File: tests/test_test.py [hunk 19]
```diff
@@ -135,8 +142,8 @@ def test_get_label_set(targets, exclude_label, expected):
     assert set(expected) == set(actual)
 
 
-async def test_interpreter_passed_to_agent(
```
Comment (wochinge) [inline tests/test_test.py:138]: @tmbo Why did you remove this test?

File: rasa/importers/importer.py [hunk 20]
```diff
@@ -20,29 +19,26 @@ async def get_domain(self) -> Domain:
...
 
         Returns:
-            Loaded ``Domain``.
+            Loaded `Domain`.
         """
         raise NotImplementedError()
 
     async def get_stories(
         self,
-        interpreter: "NaturalLanguageInterpreter" = RegexInterpreter(),
```
Comment (wochinge) [inline rasa/importers/importer.py:29]: We are changing an interface here which might be re-implemented by a community member. Can we please add a removal changelog?
